lms/djangoapps/shoppingcart/pdf.py

```diff
@@ -0,0 +1,448 @@
+"""
+Template for PDF Receipt/Invoice Generation
+"""
+from PIL import Image
+from reportlab.lib import colors
+from django.conf import settings
+from django.utils.translation import ugettext as _
+from reportlab.pdfgen.canvas import Canvas
+from reportlab.lib.pagesizes import letter
+from reportlab.lib.units import mm
+from reportlab.lib.styles import getSampleStyleSheet
+from reportlab.platypus import Paragraph
+from reportlab.platypus.tables import Table, TableStyle
+from microsite_configuration import microsite
+from xmodule.modulestore.django import ModuleI18nService
+
+
+class NumberedCanvas(Canvas):  # pylint: disable=abstract-method
+    """
+    Canvas child class with auto page-numbering.
+    """
+    def __init__(self, *args, **kwargs):
+        """
+            __init__
+        """
+        Canvas.__init__(self, *args, **kwargs)
+        self._saved_page_states = []
+
+    def insert_page_break(self):
+        """
+        Starts a new page.
+        """
+        self._saved_page_states.append(dict(self.__dict__))
+        self._startPage()
+
+    def current_page_count(self):
+        """
+        Returns the page count in the current pdf document.
+        """
+        return len(self._saved_page_states) + 1
+
+    def save(self):
+        """
+            Adds page numbering to each page (page x of y)
+        """
+        num_pages = len(self._saved_page_states)
+        for state in self._saved_page_states:
+            self.__dict__.update(state)
```

Comment:
Not really sure what's going on in there. Just a faint idea though... Copied the code from here:

http://code.activestate.com/recipes/576832/
which is an improved version of :
http://code.activestate.com/recipes/546511/

and http://stackoverflow.com/questions/1087495/reportlab-page-x-of-y-numberedcanvas-and-images

The limitation exists in reportlab which doesn't allow us to modify the page once it's finished. Since we are not sure how many pages the finished pdf will have we cannot put a page x of y at the bottom. The other solution is to make a pdf, then make another copy just for this "page x of y" string. This workaround was mentioned at different places and we just stuck it in our code without paying much attention to what it does. Just improved the naming of the function names and added a few helpers.